### PR TITLE
[JENKINS-57988] Add Java 8 streams and remove hashmap

### PR DESCRIPTION
**Pull request by baymac**
I have removed HashMap in this PR. Although the problem that UI not refreshing at runtime still exist. This PR is an alternative of #15. So only one of them gets merged.

### CODE REVIEW

**Comment by baymac** [comment]
I am also unable to use environment variable to populate the token field in `jenkins.yaml`

For example,
I added a global variable with the key: `GITLAB_PERSONAL_ACCESS_TOKEN` and value: `<token>`. And my `jenkins.yaml` looks like:

```yaml
credentials:
  system:
    domainCredentials:
      - credentials:
          - gitlabPersonalAccessToken:
              scope: SYSTEM
              id: "i<3GitLab"
              token: ${GITLAB_PERSONAL_ACCESS_TOKEN}

unclassified:
  gitLabServers:
    servers:
      - credentialsId: "i<3GitLab"
        manageHooks: true
        name: "gitlab.com"
        serverUrl: "https://gitlab.com"
```

After reloading the configuration, when I test the GitLab Server connection, it doesn't work. The token is valid, I tried replacing the env variable with the plain text token in `jenkins.yaml` it works. 

Also when I go to my Jenkins instance systeminfo (http://localhost:8002/jenkins/systemInfo), then I can't see the global variable (GITLAB_PERSONAL_ACCESS_TOKEN) I added previously.

**Comment by markjacksonfishing** [comment]
> I am also unable to use environment variable to populate the token field in `jenkins.yaml`
> 
> For example,
> I added a global variable with the key: `GITLAB_PERSONAL_ACCESS_TOKEN` and value: `<token>`. And my `jenkins.yaml` looks like:
> 
> ```yaml
> credentials:
>   system:
>     domainCredentials:
>       - credentials:
>           - gitlabPersonalAccessToken:
>               scope: SYSTEM
>               id: "i<3GitLab"
>               token: ${GITLAB_PERSONAL_ACCESS_TOKEN}
> 
> unclassified:
>   gitLabServers:
>     servers:
>       - credentialsId: "i<3GitLab"
>         manageHooks: true
>         name: "gitlab.com"
>         serverUrl: "https://gitlab.com"
> ```
> 
> After reloading the configuration, when I test the GitLab Server connection, it doesn't work. The token is valid, I tried replacing the env variable with the plain text token in `jenkins.yaml` it works.
> 
> Also when I go to my Jenkins instance systeminfo (http://localhost:8002/jenkins/systemInfo), then I can't see the global variable (GITLAB_PERSONAL_ACCESS_TOKEN) I added previously.

Hmm, that is interesting. Perhaps the refresh expects a new token? If possible can we track this via an issue as well as a jira ticket?

**Comment by baymac** [comment]
Yes. I will add a ticket for this.

**Comment by jetersen** [comment]
Where did you define this environment variable?

**Comment by jetersen** [comment]
Also has to be in quotes `token: "${GITLAB_PERSONAL_ACCESS_TOKEN}"`

**Comment by baymac** [comment]
> Also has to be in quotes `token: "${GITLAB_PERSONAL_ACCESS_TOKEN}"`

Tried that as well.

**Comment by baymac** [comment]
> Where did you define this environment variable?

In the Global Configuration
Screenshot from 2019-06-12 19-09-28

**Comment by jetersen** [comment]
That's not how JCasC works :cry: 
JCasC can definitely define global properties. but it does not read from global properties configuration.

**Comment by baymac** [comment]
Then how do I add env variable?

**Comment by jetersen** [comment]
you have to define the secrets as part of spinning up your Jenkins instance.
ie. environment variables, if your using `mvn hpi:run` you can do `GITLAB_PERSONAL_ACCESS_TOKEN="MYTOKEN" mvn hpi:run`
or inside IntelliJ run configuration, you can add the environment variable

**Comment by baymac** [comment]
Oh :+1:

**Comment by baymac** [comment]
So what is the use of global variables?

**Comment by baymac** [comment]
Do you have any idea of a fix for the UI part that is not reflecting the state of the list of `servers`?

**Comment by jetersen** [comment]
> Do you any idea of a fix for the UI part that is not reflecting the state of the list of `servers`?

Yes, but it is easier to show in code :sweat:

> So what is the use of global ~variables~ properties?

Global properties can be used to define several things, amongst there are environment variables.
These can be used to configure something globally that can be accessed inside builds and agents and usually also plugins but that would defeat the purpose of a secret :sweat:

**Comment by baymac** [comment]
> > Do you any idea of a fix for the UI part that is not reflecting the state of the list of `servers`?
> 
> Yes, but it is easier to show in code 

So when you have time can you commit to this branch?

**Comment by jetersen** [comment]
I added my commit with an explain: https://github.com/baymac/gitlab-branch-source-plugin/pull/16/commits/333daff8bcf0a0a6ec39a42885b4bebcca2d218e

**Comment by baymac** [comment]
> I added my commit with an explain: [333daff](https://github.com/baymac/gitlab-branch-source-plugin/commit/333daff8bcf0a0a6ec39a42885b4bebcca2d218e)

This was super helpful. Thanks.

**Comment by baymac** [comment]
I think this PR is ready to be merged.

**Comment by baymac** [comment]
[JENKINS-57988](https://issues.jenkins-ci.org/browse/JENKINS-57988) also fixed in this pr by @casz

**Comment by jetersen** [comment]
This is why merge commits should never exist :sob: 
image

File: README.md [hunk 1]
```diff
@@ -238,7 +238,7 @@ Assuming plugin installation has been done already.
      
 6. Adding a Personal Access Token Credentials:
 
-   This is a manual setup. To automatically generate Personal Accees Token see next section
+   This is a manual setup. To automatically generate Personal Accees Token see next [section](https://github.com/baymac/gitlab-branch-source-plugin#creating-personal-access-token-within-jenkins).
```
Comment (jetersen) [inline README.md]: You can do relative links :wink: 
```suggestion
   This is a manual setup. To automatically generate Personal Accees Token see next [section](#creating-personal-access-token-within-jenkins).
```

Should do the trick

Comment (baymac) [reply README.md]: Cool.

Comment (jetersen) [reply README.md]: https://github.com/baymac/gitlab-branch-source-plugin/blob/8fe081304dec8afbb44842a70671720ddd36046b/README.md yup works :sweat_smile:

**Comment by baymac** [comment]
> This is why merge commits should never exist 
> image

I am not quite sure what this means? You mean to say we need to `Rebase and Merge` our PRs?

**Comment by jetersen** [comment]
Squash and merge or rebase and merge is fine, it depends on the work (does it make sense to rebase if all commits are meaningful). What it does it creates a much cleaner graph.

For instance in squash and merge you can achieve the same with using trailers.
See this commit for instance: https://github.com/jenkinsci/configuration-as-code-plugin/commit/3a67b90dec1ccc0aece636836998f137f526af2a

I know that this is a smaller project but the graph is already pretty messy :laughing: 

GitHub has a limited network graph where we can compare develop branch with JCasC master branch (You can use the mouse to drag the graph)
https://github.com/jenkinsci/configuration-as-code-plugin/network
https://github.com/baymac/gitlab-branch-source-plugin/network

If you want to visualize the git graph I suggest you grab GitKraken which is included in your GitHub student pack :wink:

**Comment by baymac** [comment]
It makes sense to write useful commits. I already have gitkraken, think I will need to use it more often. I will try to rebase and merge wherever possible and in messed up branches I will do a squash and merge.

> For instance in squash and merge you can achieve the same with using trailers.
> See this commit for instance: [jenkinsci/configuration-as-code-plugin@3a67b90](https://github.com/jenkinsci/configuration-as-code-plugin/commit/3a67b90dec1ccc0aece636836998f137f526af2a)

This also makes sense. Although it will take some time for me to understand squash and merge. 

So what we do with the 2 pending PRs? Rebase and merge?

**Comment by jetersen** [comment]
> So what we do with the 2 pending PRs? Rebase and merge?

Feel free to rebase and merge

> Although it will take some time for me to understand squash and merge.

Squash and merge are pretty basic.

Leaves the pull requests history intact which is a huge benefit when you live on GitHub and rely on the pull request.

Takes all the commits in the PR and squash it into a single commit and then merge that into master as a single commit. The difference is that is done as a commit and not as a merge commit.

Squash and merge also allows you to edit the commit message before merging the commit in.

GitHub has excellent articles on these topics: https://help.github.com/en/articles/about-pull-request-merges#squash-and-merge-your-pull-request-commits

**Comment by baymac** [comment]
Thanks will look into it while I squash merge this pr. :sweat_smile:

**Pull request merged**